[AutoAC] Backward Pass Aware AC - changes to partitioner to acommodate SOLVER as a callable

Chillee opened the pull request:
Summary: making it so that the config can pass `config.activation_memory_budget_solver` as a callable method and then that callable is invoked to determine the set of saved/recomputed nodes.

Test Plan: tbd

Reviewed By: Chillee, basilwong

Differential Revision: D63714905

basilwong commented on torch/_functorch/partitioners.py:
```diff
@@ -1517,6 +1520,11 @@ def _optimize_runtime_with_given_memory(
+        saved_node_idx, recomp_node_idx = SOLVER(
+            memory, joint_graph, max_memory, node_info, all_recomputable_banned_nodes
+        )
+        return (0.0, saved_node_idx, recomp_node_idx)
```
Nice, will rebase changes onto tuple

paryxyt committed:
[AutoAC] Backward Pass Aware AC - changes to partitioner to acommodate SOLVER as a callable (#137314)

Summary:

making it so that the config can pass `config.activation_memory_budget_solver` as a callable method and then that callable is invoked to determine the set of saved/recomputed nodes.

Test Plan: tbd

Reviewed By: Chillee, basilwong

Differential Revision: D63714905

huydhn commented:
@pytorchbot revert -m 'The failure shows up in trunk' -c ignoredsignal

functorch/test_ac.py::MemoryBudgetTest::test_attention_vs_linear [GH job link](https://github.com/pytorch/pytorch/actions/runs/11246051611/job/31270759700) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/932b9945c0bc61a11a7db2f52c974cf283d5a2ed)
